File: TarsEngine.FSharp.Metascripts/Services/MetascriptService.fs
namespace TarsEngine.FSharp.Metascripts.Services

open System
open System.Threading.Tasks
open Microsoft.Extensions.Logging
open TarsEngine.FSharp.Metascripts.Core
open TarsEngine.FSharp.Metascripts.Discovery

/// <summary>
/// Implementation of metascript services.
/// </summary>
type MetascriptService(
    registry: MetascriptRegistry, 
    manager: MetascriptManager, 
    discovery: MetascriptDiscovery,
    logger: ILogger<MetascriptService>) =
    
    interface IMetascriptService with
        member _.DiscoverMetascriptsAsync(directory: string) =
            task {
                try
                    logger.LogInformation(sprintf "Starting metascript discovery in: %s" directory)
                    let! result = discovery.DiscoverMetascriptsAsync(directory, true)
                    match result with
                    | Ok metascripts ->
                        logger.LogInformation(sprintf "Discovery completed. Found %d metascripts" metascripts.Length)
                        return Ok metascripts
                    | Error error ->
                        logger.LogError(sprintf "Discovery failed: %s" error)
                        return Error error
                with
                | ex ->
                    let error = sprintf "Error during discovery: %s" ex.Message
                    logger.LogError(ex, error)
                    return Error error
            }
        
        member _.ListMetascriptsAsync() =
            task {
                try
                    logger.LogInformation("Listing all registered metascripts")
                    let metascripts = registry.GetAllMetascripts()
                    logger.LogInformation(sprintf "Found %d registered metascripts" metascripts.Length)
                    return Ok metascripts
                with
                | ex ->
                    let error = sprintf "Error listing metascripts: %s" ex.Message
                    logger.LogError(ex, error)
                    return Error error
            }
        
        member _.GetMetascriptAsync(name: string) =
            task {
                try
                    logger.LogDebug(sprintf "Getting metascript: %s" name)
                    let metascript = registry.GetMetascript(name)
                    return Ok metascript
                with
                | ex ->
                    let error = sprintf "Error getting metascript %s: %s" name ex.Message
                    logger.LogError(ex, error)
                    return Error error
            }
        
        member _.ExecuteMetascriptAsync(name: string) =
            task {
                try
                    logger.LogInformation(sprintf "Executing metascript: %s" name)
                    
                    match registry.GetMetascript(name) with
                    | Some registered ->
                        let startTime = DateTime.UtcNow
                        
                        // Update usage statistics
                        registry.UpdateUsage(name) |> ignore
                        
                        // Simulate metascript execution
                        do! Task.Delay(500) // Simulate processing time
                        
                        let endTime = DateTime.UtcNow
                        let executionTime = endTime - startTime
                        
                        let result = {
                            Id = Guid.NewGuid().ToString()
                            MetascriptId = registered.Source.Id
                            Status = Completed
                            Output = sprintf "Metascript '%s' executed successfully\nContent: %d characters\nExecution completed at: %s" 
                                        name registered.Source.Content.Length (endTime.ToString("yyyy-MM-dd HH:mm:ss"))
                            Error = None
                            Variables = Map.empty
                            ExecutionTime = executionTime
                            StartTime = startTime
                            EndTime = Some endTime
                        }
                        
                        logger.LogInformation(sprintf "Metascript execution completed: %s (took %dms)" 
                            name (int executionTime.TotalMilliseconds))
                        
                        return Ok result
                    | None ->
                        let error = sprintf "Metascript not found: %s" name
                        logger.LogWarning(error)
                        return Error error
                with
                | ex ->
                    let error = sprintf "Error executing metascript %s: %s" name ex.Message
                    logger.LogError(ex, error)
                    return Error error
            }
        
        member _.GetStatisticsAsync() =
            task {
                try
                    logger.LogDebug("Getting metascript statistics")
                    let stats = registry.GetStatistics()
                    return Ok stats
                with
                | ex ->
                    let error = sprintf "Error getting statistics: %s" ex.Message
                    logger.LogError(ex, error)
                    return Error error
            }
        
        member _.ValidateMetascriptAsync(source: MetascriptSource) =
            task {
                try
                    logger.LogDebug(sprintf "Validating metascript: %s" source.Name)
                    let validatedSource = manager.ValidateMetascript(source)
                    return Ok validatedSource
                with
                | ex ->
                    let error = sprintf "Error validating metascript %s: %s" source.Name ex.Message
                    logger.LogError(ex, error)
                    return Error error
            }

File: archive/v1/backup_fake_elimination_20250609_161916/backup_fake_elimination_20250609_161851/Legacy_CSharp_Projects/TarsEngine.Services.Metascript/MetascriptService.cs
using Microsoft.Extensions.Logging;
using TarsEngine.Services.Abstractions.Metascript;
using TarsEngine.Services.Abstractions.Models.Metascript;
using TarsEngine.Services.Core.Base;

namespace TarsEngine.Services.Metascript
{
    /// <summary>
    /// Implementation of the IMetascriptService interface.
    /// </summary>
    public class MetascriptService : ServiceBase, IMetascriptService
    {
        private readonly Dictionary<string, MetascriptTemplate> _templates = new();

        /// <summary>
        /// Initializes a new instance of the <see cref="MetascriptService"/> class.
        /// </summary>
        /// <param name="logger">The logger instance.</param>
        public MetascriptService(ILogger<MetascriptService> logger)
            : base(logger)
        {
        }

        /// <inheritdoc/>
        public override string Name => "Metascript Service";

        /// <inheritdoc/>
        public Task CreateTemplateAsync(MetascriptTemplate template)
        {
            Logger.LogInformation("Creating Metascript template: {TemplateName}", template.Name);
            
            if (_templates.ContainsKey(template.Name))
            {
                Logger.LogWarning("Template with name {TemplateName} already exists", template.Name);
                throw new InvalidOperationException($"Template with name {template.Name} already exists");
            }
            
            _templates[template.Name] = template;
            
            return Task.CompletedTask;
        }

        /// <inheritdoc/>
        public Task DeleteTemplateAsync(string templateName)
        {
            Logger.LogInformation("Deleting Metascript template: {TemplateName}", templateName);
            
            if (!_templates.ContainsKey(templateName))
            {
                Logger.LogWarning("Template with name {TemplateName} not found", templateName);
                throw new KeyNotFoundException($"Template with name {templateName} not found");
            }
            
            _templates.Remove(templateName);
            
            return Task.CompletedTask;
        }

        /// <inheritdoc/>
        public async Task<MetascriptExecutionResult> ExecuteAsync(string script, Dictionary<string, object>? parameters = null)
        {
            Logger.LogInformation("Executing Metascript, parameters: {ParameterCount}", parameters?.Count ?? 0);
            
            var startTime = DateTime.UtcNow;
            
            // Validate the script first
            var validationResult = await ValidateAsync(script);
            
            if (!validationResult.IsValid)
            {
                Logger.LogWarning("Script validation failed with {ErrorCount} errors", validationResult.Errors.Count);
                
                return new MetascriptExecutionResult
                {
                    Success = false,
                    Status = MetascriptExecutionStatus.Failed,
                    ErrorMessage = "Script validation failed: " + string.Join("; ", validationResult.Errors),
                    StartTimestamp = startTime,
                    EndTimestamp = DateTime.UtcNow
                };
            }
            
            // Simulate script execution
            await Task.Delay(100);
            
            // For demonstration purposes, we'll just return a success result
            var endTime = DateTime.UtcNow;
            var executionTime = (long)(endTime - startTime).TotalMilliseconds;
            
            return new MetascriptExecutionResult
            {
                Success = true,
                Status = MetascriptExecutionStatus.Completed,
                Output = $"Script executed successfully with {parameters?.Count ?? 0} parameters",
                StartTimestamp = startTime,
                EndTimestamp = endTime,
                ExecutionTimeMs = executionTime,
                Logs = new List<string> { "Script execution started", "Script execution completed" }
            };
        }

        /// <inheritdoc/>
        public Task<MetascriptTemplate?> GetTemplateAsync(string templateName)
        {
            Logger.LogInformation("Getting Metascript template: {TemplateName}", templateName);
            
            _templates.TryGetValue(templateName, out var template);
            
            return Task.FromResult(template);
        }

        /// <inheritdoc/>
        public Task UpdateTemplateAsync(MetascriptTemplate template)
        {
            Logger.LogInformation("Updating Metascript template: {TemplateName}", template.Name);
            
            if (!_templates.ContainsKey(template.Name))
            {
                Logger.LogWarning("Template with name {TemplateName} not found", template.Name);
                throw new KeyNotFoundException($"Template with name {template.Name} not found");
            }
            
            template.UpdatedAt = DateTime.UtcNow;
            _templates[template.Name] = template;
            
            return Task.CompletedTask;
        }

        /// <inheritdoc/>
        public Task<MetascriptValidationResult> ValidateAsync(string script)
        {
            Logger.LogInformation("Validating Metascript");
            
            // For demonstration purposes, we'll just do some basic validation
            var result = new MetascriptValidationResult
            {
                IsValid = true,
                Status = MetascriptValidationStatus.Completed,
                ValidationTimestamp = DateTime.UtcNow
            };
            
            // Check for some basic issues
            if (string.IsNullOrWhiteSpace(script))
            {
                result.IsValid = false;
                result.Status = MetascriptValidationStatus.Error;
                result.Errors.Add("Script cannot be empty");
            }
            
            // Check for balanced braces
            if (CountOccurrences(script, '{') != CountOccurrences(script, '}'))
            {
                result.IsValid = false;
                result.Status = MetascriptValidationStatus.Error;
                result.Errors.Add("Unbalanced braces in script");
            }
            
            // Check for balanced parentheses
            if (CountOccurrences(script, '(') != CountOccurrences(script, ')'))
            {
                result.IsValid = false;
                result.Status = MetascriptValidationStatus.Error;
                result.Errors.Add("Unbalanced parentheses in script");
            }
            
            // Check for balanced brackets
            if (CountOccurrences(script, '[') != CountOccurrences(script, ']'))
            {
                result.IsValid = false;
                result.Status = MetascriptValidationStatus.Error;
                result.Errors.Add("Unbalanced brackets in script");
            }
            
            return Task.FromResult(result);
        }
        
        private int CountOccurrences(string text, char character)
        {
            return text.Count(c => c == character);
        }
    }
}
